### PR TITLE
[CSClosure] Fix property wrapper handling

### LINKED ISSUE #61017: [Regression] 5.7 – "cannot reference invalid declaration" error when accessing property wrapper inside closure
**Describe the bug**
Accessing a property wrapper using the underscore notation in a closure does not compile with Swift 5.7 (Xcode 14 RC).

This code reproduces the issue:

```swift
@propertyWrapper
struct Wrapper {
    var wrappedValue: Int
    
    init(wrappedValue: Int) {
        self.wrappedValue = wrappedValue
    }
}

class Foo {
    let bar: () -> Void = {
        @Wrapper var wrapped = 1
        
        let wrapper = _wrapped // compiles on 5.6, errors out on 5.7
    }
}

```

**Steps To Reproduce**
1. Add the code above to a Wrapper.swift file
2. Compile the file with the Swift 5.7 compiler: `swiftc Wrapper.swift`
3. The following compilation error occurs: 
```
❯ swiftc Wrapper.swift
Wrapper.swift:14:23: error: cannot reference invalid declaration '_wrapped'
        let wrapper = _wrapped // compiles on 5.6, errors out on 5.7
                      ^
Wrapper.swift:12:22: note: '_wrapped' declared here
        @Wrapper var wrapped = 1
```

**Expected behavior**
Accessing the property wrapper using the underscore notation code should compile, as it does on Swift 5.6.

**Environment**
 - Xcode 14 RC toolchain

**Additional context**
The same code produces a SILGen crash with the 5.7 nightly toolchain: https://godbolt.org/z/o1qGnov7o

```
unsupported collection upcast kind
UNREACHABLE executed at /home/build-user/swift/lib/SILGen/SILGenPoly.cpp:523!
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the project and the crash backtrace.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/swift-5.7/usr/bin/swift-frontend -frontend -S -primary-file <source> -target x86_64-unknown-linux-gnu -disable-objc-interop -g -new-driver-path /opt/compiler-explorer/swift-5.7/usr/bin/swift-driver -empty-abi-descriptor -resource-dir /opt/compiler-explorer/swift-5.7/usr/lib/swift -enable-anonymous-context-mangled-names -Xllvm --x86-asm-syntax=intel -module-name output -o /app/output.s
1.	Swift version 5.7.1-dev (LLVM 7e31b78ab3255fa, Swift 1e80674d67f5505)
2.	Compiling with the current language version
3.	While evaluating request ASTLoweringRequest(Lowering AST to SIL for file "<source>")
4.	While silgen emitStoredPropertyInitialization SIL function "@$s6output3FooC3baryycvpfi".
 for expression at [<source>:11:27 - line:15:5] RangeText="{
        @Wrapper var wrapped = 1
        
        let wrapper = _wrapped // works on 5.6, crashes on 5.7
    "
5.	While silgen closureexpr SIL function "@$s6output3FooC3baryycvpfiyycfU_".
 for expression at [<source>:11:27 - line:15:5] RangeText="{
        @Wrapper var wrapped = 1
        
        let wrapper = _wrapped // works on 5.6, crashes on 5.7
    "

...

Compiler returned: 139
```

### LINKED ISSUE #61024: [Regression][5.7] "unable to infer closure type in the current context" when passing a member access expression as property wrapper argument inside closure
**Describe the bug**
Passing a member access expression (such as `SomeEnum.someCase.rawValue`, `SomeType().someProp`, etc.) as an argument to a property wrapper located inside a closure causes a compilation error on Swift 5.7, whereas it compiles correctly on Swift 5.6. 

The following code reproduces the issue:

```swift
@propertyWrapper
struct Wrapper {    
    var wrappedValue: Int
    let argument: String
    
    init(wrappedValue: Int, argument: String) {
        self.wrappedValue = wrappedValue
        self.argument = argument
    }
}

class Foo {
    let bar: () -> Void = {
        @Wrapper(argument: Baz.someCase.rawValue) // works on 5.6, errors out on 5.7
        var wrapped = 1
    }
}

enum Baz: String {
    case someCase
}
```

**Steps To Reproduce**
1. Add the code above to a Wrapper.swift file
2. Compile the file with the Swift 5.7 compiler: `$ swiftc Wrapper.swift`
3. The following compilation error occurs:
```
❯ swiftc Wrapper.swift
Wrapper.swift:13:27: error: unable to infer closure type in the current context
    let bar: () -> Void = {
                          ^
```

**Expected behavior**
Passing a member access expression that returns a type that is compatible with the property wrapper's parameter that it's being passed to should compile correctly when located inside a closure, as it does on Swift 5.6.

**Environment**
 - Xcode 14 RC toolchain (Swift 5.7)

**Additional context**
The code snippet above demonstrates the issue using an enum with a raw value, but it seems the issue happens with other member access expressions as well, such as:

```swift
class Foo {
    let bar: () -> Void = {
        @Wrapper(argument: Baz().someProp) // works on 5.6, errors out on 5.7
        var wrapped = 1
    }
}

struct Baz {
    let someProp: String = "someArg"
}
```
Also, this issue only seems to happen in closures, changing the closure declaration to a method with the same signature compiles correctly:

```swift
class Foo {
    func bar() {
        @Wrapper(argument: Baz.someCase.rawValue) // works on 5.6 and 5.7
        var wrapped = 1
    }
}
```

5.6 vs 5.7 compiler comparison here: https://godbolt.org/z/shEs56dhM

### PR DESCRIPTION
**Pull request by xedin**
- Always pre-check `init` targets of wrapped vars

If a variable with attached property wrapper has an initializer
expression it could be modified by implicit wrapper application,
if there is no initializer - one would be synthesized by the
compiler (without arguments). In both cases target has to be
pre-checked before constraints are generated for it.

- Use correct type for implicit wrapper variables

Using `computeWrappedValueType` is incorrect because
that return a type of the wrapped variable and not
the *wrapper* variable (one that starts with `_`).

Resolves: https://github.com/apple/swift/issues/61024
Resolves: https://github.com/apple/swift/issues/61017

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Review by hborla — APPROVED** [review]

File: lib/Sema/CSClosure.cpp [hunk 1]
```diff
@@ -530,18 +530,13 @@ class SyntacticElementConstraintGenerator
...
-          /*bindPatternVarsOneWay=*/false);
-    }
-
-    // If there was no initializer, there could be one from a property
-    // wrapper which has to be pre-checked before use. This is not a
-    // problem in top-level code because pattern bindings go through
-    // `typeCheckExpression` which does pre-check automatically and
-    // result builders do not allow declaring local wrapped variables.
+    // A property wrapper initializer (either user-defined
+    // for a synthesized one) has to be pre-checked before use.
```
Comment (theblixguy) [inline lib/Sema/CSClosure.cpp:534]: Looks like there's a small typo:
```suggestion
    // or a synthesized one) has to be pre-checked before use.
```

Comment (xedin) [reply lib/Sema/CSClosure.cpp]: Thanks, fixed!

**Comment by xedin** [comment]
@swift-ci please test

File: test/expr/closure/multi_statement.swift [hunk 2]
```diff
@@ -566,3 +566,63 @@ struct Test {
...
+  class Test {
+    let bar: () -> Void = {
+      @Wrapper var wrapped = 1
+      let wrapper: Wrapper = _wrapped // Ok
+    }
+  }
+}
+
+https://github.com/apple/swift/issues/61024
+do {
```
Comment (AnthonyLatsis) [inline test/expr/closure/multi_statement.swift:590]: Took me a while to realize why this link compiles haha (statement label + comment).

**Pull request merged**